Add validation of scanf format and arguments

Luhrel opened the pull request:
@WalterBright asked for it in [the forums](https://forum.dlang.org/thread/r35188$1u2o$1@digitalmars.com), so here it is.

Highly inspired by #10812.

RazvanN7 commented on src/dmd/chkscanf.d:
```diff
@@ -0,0 +1,641 @@
+import dmd.mtype;
+import dmd.target;
+
+import std.ascii : isDigit;
```
We do not use the standard library in the compiler to avoid the chicken and egg problem. You will most likely find a similar function in either the root directory or most definitely in the lexer class.

WalterBright replied:
Besides, the implementation is *so trivial* it just isn't worth the dependency.

Luhrel replied:
Fixed.

RazvanN7 commented on src/dmd/chkscanf.d:
```diff
@@ -0,0 +1,641 @@
+import dmd.target;
+
+import std.ascii : isDigit;
+import std.string : isNumeric;
```
ditto.

WalterBright replied:
ditto

RazvanN7 reviewed (changes requested):
Most of the logic here seems like a duplicate of `chkprintf.d`. Is there any difference between the mirrored functions? Even if there is, it should be minimal. I suggest that `chkprintf` file is renamed to `chkformat` and try to reuse the code as much as possible. For example: the `parseFormatSpecifier` function should be identical for both situations, there is no need in duplicating it; even if there are some differences, those should be special cased with if-else branches. Another example is the `enum Format` declaration: 2 declarations of the same thing.

WalterBright commented on src/dmd/chkscanf.d:
```diff
@@ -0,0 +1,641 @@
+    {
+        while (isDigit(format[i]))
+        {
+            width ~= format[i];
```
There is no reason to store this as a string, or even store it at all. The only purpose here for the field width is to diagnose an invalid format when used with the `n` specifier.

Luhrel replied:
`width` removed, as well as the `out` arg. Should we check for any invalid format with `n`, or is it too strict ?

WalterBright commented on src/dmd/chkscanf.d:
```diff
@@ -0,0 +1,639 @@
+        asteriskSubSpec = true;
+    }
+
+    // width sub-specifier
```
Please call this `fieldWidth` to match the C99 terminology.

Luhrel committed:
Implements chkscanf.d

thewilsonator commented:
Are `sscanf` and `fscanf` supported? (also for that matter `sprtinf` and `fprintf`?)

Luhrel committed:
Implements chkformat.d

Now `chkscanf` and `chkprintf` use the same base function for parsing a
generic `Format`.

Luhrel committed:
Add validation for sscanf, fscanf, sprintf and fprintf

Luhrel commented:
> Most of the logic here seems like a duplicate of `chkprintf.d`. Is there any difference between the mirrored functions? Even if there is, it should be minimal. I suggest that `chkprintf` file is renamed to `chkformat` and try to reuse the code as much as possible. For example: the `parseFormatSpecifier` function should be identical for both situations, there is no need in duplicating it; even if there are some differences, those should be special cased with if-else branches. Another example is the `enum Format` declaration: 2 declarations of the same thing.

Done.

> Are `sscanf` and `fscanf` supported? (also for that matter `sprtinf` and `fprintf`?)

Now, yes.

WalterBright commented on src/dmd/chkprintf.d:
```diff
@@ -397,115 +383,34 @@ Format parseFormatSpecifier(scope const char[] format, ref size_t idx,
 
         default:
-            return error();
+            break;
```
An undefined specifier should be an error. Please check all "undefined behavior" notations in the C99 spec and make sure they're all covered in the logic and the test cases.

Luhrel replied:
Now it's the `parseGenericFormatSpecifier` function (chkformat.d) that checks for undefined behavior.
This `switch` now just checks if there is an incompatible type and specifier (i.e `%#d`). If so, then it returns an error.

Luhrel replied:
Now it's the `parseGenericFormatSpecifier` function (chkformat.d) that checks for undefined behavior.
This `switch` now just checks if there is an incompatible type and specifier (i.e `%#d`). If so, then it returns an error.

WalterBright replied:
If the `default:` case is unreachable, the code for it should be `assert(0);` not `break;`

thewilsonator commented on changelog/chkscanf.md:
````diff
@@ -0,0 +1,48 @@
+should be replaced with:
+```
+int i;
+printf("%d\n", &i;
````
this should be `scanf`

thewilsonator commented on changelog/chkscanf.md:
```diff
@@ -0,0 +1,48 @@
+size_t s;
+scanf("%zd\n", &s);
+ulong u;
+printf("%llu%*c\n", u);
```
ditto

thewilsonator reviewed (changes requested):
Also the changelog should mention that the file and string variants of scanf are also validated. This changelog entry should be merged with the printf one (if that hasn't already been released). It should also mention that fprintf and sprintf are also validated.

thewilsonator commented on src/dmd/expressionsem.d:
```diff
@@ -2150,13 +2151,33 @@ private bool functionParameters(const ref Loc loc, Scope* sc,
-            if (checkPrintfFormat(se.loc, se.peekString(), (*arguments)[1 .. nargs]))
+            auto se = (*arguments)[0].isStringExp();
+            if (se && checkPrintfFormat(se.loc, se.peekString(), (*arguments)[1 .. nargs]))
+                err = true;
```
the body of all of these are almost identical. Please write it as 
```d
int paramOffset = 0; // or some better name
if (fd.ident == Id.printf || fd.ident == Id.scanf)
    paramOffset = 1;
else if (fd.ident == Id.sprintf || fd.ident == Id.fprintf ||
            fd.ident == Id.sscanf || fd.ident == Id.fscanf)
    paramOffset = 2;
if (nparams >= paramOffset && paramOffset)
{
    auto se = (*arguments)[paramOffset-1].isStringExp();
    if (se && checkPrintfFormat(se.loc, se.peekString(), 
                                                    (*arguments)[paramOffset .. nargs]))
         err = true;
    }
}
```

Luhrel replied:
`printf` and `scanf` do not use the same function. Going to write something similar.

thewilsonator commented on src/dmd/chkscanf.d:
```diff
@@ -0,0 +1,498 @@
+    }
+
+    // fieldWidth
+    {
```
please remove this pair of braces. Keep the comment.

thewilsonator commented on src/build.d:
```diff
@@ -1168,7 +1168,7 @@ auto sourceFiles()
             parse.d parsetimevisitor.d permissivevisitor.d printast.d safe.d sapply.d scanelf.d scanmach.d
             scanmscoff.d scanomf.d semantic2.d semantic3.d sideeffect.d statement.d statement_rewrite_walker.d
             statementsem.d staticassert.d staticcond.d target.d templateparamsem.d traits.d
-            transitivevisitor.d typesem.d typinf.d utils.d visitor.d foreachvar.d chkprintf.d
```
Please no dot add more files. Please merge them into one file called `chkformat.d`. We have more than enough files on the root directory already.

Luhrel committed:
Update documentation

Luhrel committed:
Simplifying code as requested.

Luhrel committed:
Merge chkscanf and chkprintf into chkformat

Luhrel committed:
Fix runnable/test20.d printf args

thewilsonator commented:
This PR seems to have some issues with `real`. See the autotester for details e.g. https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=3964672&isPull=true

Luhrel committed:
Merge documentation

andy-k commented on changelog/chkscanf.md:
````diff
@@ -0,0 +1,50 @@
+should be replaced with:
+```
+int i;
+scanf("%d\n", &i;
````
missing `)`

andy-k commented on changelog/chkscanf.md:
```diff
@@ -0,0 +1,50 @@
+size_t s;
+scanf("%zd\n", &s);
+ulong u;
+scanf("%llu%*c\n", u);
```
should this (and above) be `&u`?

andy-k commented on changelog/chkscanf.md:
````diff
@@ -0,0 +1,50 @@
+```
+```
+const format = "%k\n";
+scanf(format.ptr, value);  // no error
````
should these be `&value`?

Luhrel committed:
Little cleanup

WalterBright commented:
The errors in the test suite are about druntime's use of `sprintf`. This  is why `sprintf` support should not have been lumped into this PR, as this PR is putatively about `scanf`.

I recommend removing `fprintf` and `sprintf` from this PR and putting them in a separate PR.

Luhrel committed:
Remove validation for sprintf and fprintf

Luhrel commented:
Done. @WalterBright Maybe you can merge this now.
